File: Dopamine.Controls/SearchBox.vb
Public Class SearchBox
    Inherits TextBox

#Region "Variables"
    Private mSearchIconCross As Viewbox
    Private mSearchIconGlass As Viewbox
    Private mSearchButton As Label
#End Region

#Region "Properties"
    Public Property HasText As Boolean
        Get
            Return CBool(GetValue(HasTextProperty))
        End Get

        Set(ByVal value As Boolean)
            SetValue(HasTextProperty, value)
        End Set
    End Property

    Public Property HasFocus As Boolean
        Get
            Return CBool(GetValue(HasFocusProperty))
        End Get

        Set(ByVal value As Boolean)
            SetValue(HasFocusProperty, value)
        End Set
    End Property

    Public Property Accent As Brush
        Get
            Return CType(GetValue(AccentProperty), Brush)
        End Get

        Set(ByVal value As Brush)
            SetValue(AccentProperty, value)
        End Set
    End Property

    Public Property SearchGlassForeground As Brush
        Get
            Return CType(GetValue(SearchGlassForegroundProperty), Brush)
        End Get

        Set(ByVal value As Brush)
            SetValue(SearchGlassForegroundProperty, value)
        End Set
    End Property
#End Region

#Region "Dependency Properties"
    Public Shared ReadOnly HasTextProperty As DependencyProperty = DependencyProperty.Register("HasText", GetType(Boolean), GetType(SearchBox), New PropertyMetadata(False))
    Public Shared ReadOnly HasFocusProperty As DependencyProperty = DependencyProperty.Register("HasFocus", GetType(Boolean), GetType(SearchBox), New PropertyMetadata(False))
    Public Shared ReadOnly AccentProperty As DependencyProperty = DependencyProperty.Register("Accent", GetType(Brush), GetType(SearchBox), New PropertyMetadata(Nothing))
    Public Shared ReadOnly SearchGlassForegroundProperty As DependencyProperty = DependencyProperty.Register("SearchGlassForeground", GetType(Brush), GetType(SearchBox), New PropertyMetadata(Nothing))
#End Region
#Region "Construction"
    Shared Sub New()
        'This OverrideMetadata call tells the system that this element wants to provide a style that is different than its base class.
        'This style is defined in themes\generic.xaml
        DefaultStyleKeyProperty.OverrideMetadata(GetType(SearchBox), New FrameworkPropertyMetadata(GetType(SearchBox)))
    End Sub
#End Region

#Region "Functions"
    Public Overrides Sub OnApplyTemplate()
        MyBase.OnApplyTemplate()

        Me.mSearchIconCross = DirectCast(GetTemplateChild("PART_SearchIconCross"), Viewbox)
        Me.mSearchIconGlass = DirectCast(GetTemplateChild("PART_SearchIconGlass"), Viewbox)
        Me.mSearchButton = DirectCast(GetTemplateChild("PART_SearchButton"), Label)

        If Me.mSearchButton IsNot Nothing Then
            AddHandler Me.mSearchButton.MouseLeftButtonUp, AddressOf SearchButton_MouseLeftButtonUphandler
        End If

        Me.SetButtonState()
    End Sub

    Private Sub SetButtonState()

        If Me.mSearchIconCross IsNot Nothing AndAlso Me.mSearchIconGlass IsNot Nothing Then

            If Me.HasText Then
                Me.mSearchIconCross.Visibility = Windows.Visibility.Visible
                Me.mSearchIconGlass.Visibility = Windows.Visibility.Collapsed
            Else
                Me.mSearchIconCross.Visibility = Windows.Visibility.Collapsed
                Me.mSearchIconGlass.Visibility = Windows.Visibility.Visible
            End If
        End If
    End Sub
#End Region

#Region "Event handlers"
    Protected Overrides Sub OnTextChanged(e As TextChangedEventArgs)
        MyBase.OnTextChanged(e)

        Me.HasText = Text.Length > 0
        Me.SetButtonState()
    End Sub

    Private Sub SearchButton_MouseLeftButtonUphandler(sender As Object, e As MouseButtonEventArgs)

        If Me.HasText Then
            Me.Text = String.Empty
        End If
        Me.SetButtonState()
    End Sub
#End Region
End Class

File: Dopamine.Common/Controls/SearchBox.cs
using System;
using System.Windows;
using System.Windows.Controls;
using System.Windows.Input;
using System.Windows.Media;

namespace Dopamine.Common.Controls
{
    public class SearchBox : TextBox
    {
        #region Variables
        private TextBlock searchIconCross;
        private TextBlock searchIconGlass;
        private Border searchBorder;
        #endregion

        #region Properties
        public bool HasText
        {
            get { return Convert.ToBoolean(GetValue(HasTextProperty)); }

            set { SetValue(HasTextProperty, value); }
        }

        public bool HasFocus
        {
            get { return Convert.ToBoolean(GetValue(HasFocusProperty)); }

            set { SetValue(HasFocusProperty, value); }
        }

        public Brush Accent
        {
            get { return (Brush)GetValue(AccentProperty); }

            set { SetValue(AccentProperty, value); }
        }

        public Brush SearchGlassForeground
        {
            get { return (Brush)GetValue(SearchGlassForegroundProperty); }

            set { SetValue(SearchGlassForegroundProperty, value); }
        }

        public Brush VisibleBackground
        {
            get { return (Brush)GetValue(VisibleBackgroundProperty); }

            set { SetValue(VisibleBackgroundProperty, value); }
        }
        #endregion

        #region Dependency Properties
        public static readonly DependencyProperty HasTextProperty = DependencyProperty.Register("HasText", typeof(bool), typeof(SearchBox), new PropertyMetadata(false));
        public static readonly DependencyProperty HasFocusProperty = DependencyProperty.Register("HasFocus", typeof(bool), typeof(SearchBox), new PropertyMetadata(false));
        public static readonly DependencyProperty AccentProperty = DependencyProperty.Register("Accent", typeof(Brush), typeof(SearchBox), new PropertyMetadata(null));
        public static readonly DependencyProperty SearchGlassForegroundProperty = DependencyProperty.Register("SearchGlassForeground", typeof(Brush), typeof(SearchBox), new PropertyMetadata(null));
        public static readonly DependencyProperty VisibleBackgroundProperty = DependencyProperty.Register("VisibleBackground", typeof(Brush), typeof(SearchBox), new PropertyMetadata(null));
        #endregion

        #region Construction
        static SearchBox()
        {
            //This OverrideMetadata call tells the system that this element wants to provide a style that is different than its base class.
            //This style is defined in themes\generic.xaml
            DefaultStyleKeyProperty.OverrideMetadata(typeof(SearchBox), new FrameworkPropertyMetadata(typeof(SearchBox)));
        }
        #endregion

        #region Functions
        public override void OnApplyTemplate()
        {
            base.OnApplyTemplate();

            this.searchIconCross = (TextBlock)GetTemplateChild("PART_SearchIconCross");
            this.searchIconGlass = (TextBlock)GetTemplateChild("PART_SearchIconGlass");
            this.searchBorder = (Border)GetTemplateChild("PART_SearchBorder");

            if (this.searchBorder != null)
            {
                this.searchBorder.MouseLeftButtonUp += SearchButton_MouseLeftButtonUphandler;
            }

            this.SetButtonState();
        }


        private void SetButtonState()
        {

            if (this.searchIconCross != null && this.searchIconGlass != null)
            {
                if (this.HasText)
                {
                    this.searchIconCross.Visibility = Visibility.Visible;
                    this.searchIconGlass.Visibility = Visibility.Collapsed;
                }
                else
                {
                    this.searchIconCross.Visibility = Visibility.Collapsed;
                    this.searchIconGlass.Visibility = Visibility.Visible;
                }
            }
        }
        #endregion

        #region Event handlers
        protected override void OnTextChanged(TextChangedEventArgs e)
        {
            base.OnTextChanged(e);
            if (string.IsNullOrWhiteSpace(this.Text)) this.Text = string.Empty; // Make sure a search cannot start by a space
            this.HasText = this.Text.Length > 0;
            this.SetButtonState();
        }

        private void SearchButton_MouseLeftButtonUphandler(object sender, MouseButtonEventArgs e)
        {
            if (this.HasText)
            {
                this.Text = string.Empty;
            }
            this.SetButtonState();
        }
        #endregion
    }
}
